scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java

```diff
@@ -0,0 +1,192 @@
+                        fieldBuilder = builder.newBuilderForField(field);
+                      }
+                      return createFieldAccessor(field, fieldBuilder);
+                    }));
+  }
+
+  @Override
+  public StructureValue valueOf(T object) {
+    return StructureValue.of(
+        object
+            .getAllFields()
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    entry -> entry.getKey().getName(),
+                    entry -> {
+                      final FieldDescriptor field = entry.getKey();
+                      @SuppressWarnings("unchecked")
+                      final AttributeValueAccessor<Object, Object> accessor =
+                          (AttributeValueAccessor<Object, Object>)
+                              fieldAccessors.get(field.getName());
+                      if (field.getJavaType().equals(JavaType.MESSAGE) && !field.isRepeated()) {
+                        StructureValue v = (StructureValue) accessor.valueOf(entry.getValue());
+                        return v.value();
+                      }
+                      return accessor.valueOf(entry.getValue());
+                    })));
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public T createFrom(StructureValue input) {
+    final Builder builder = getDefaultValue().newBuilderForType();
+
+    for (FieldDescriptor fieldDescriptor : getDefaultValue().getDescriptorForType().getFields()) {
+      final String fieldName = fieldDescriptor.getName();
+      if (input.containsKey(fieldName)) {
+        final AttributeValueAccessor<Object, Object> fieldAccessor =
+            (AttributeValueAccessor<Object, Object>) fieldAccessors.get(fieldName);
+        final Object value = prepareFieldValue(fieldDescriptor, input.get(fieldName));
+        if (fieldDescriptor.isRepeated()) {
+          final List<Object> listValues = (List<Object>) fieldAccessor.createFrom(value);
+          listValues.forEach(v -> builder.addRepeatedField(fieldDescriptor, v));
+        } else {
+          if (List.class.isAssignableFrom(value.getClass())) {
+            builder.setField(
+                fieldDescriptor, fieldAccessor.createFrom(Iterables.getLast((List<?>) value)));
+          } else {
+            builder.setField(fieldDescriptor, fieldAccessor.createFrom(value));
+          }
+        }
+      }
+    }
+
+    return (T) builder.build();
+  }
+
+  @SuppressWarnings("unchecked")
+  private Object prepareFieldValue(FieldDescriptor descriptor, Object object) {
+    if (descriptor.isRepeated()) {
+      if (List.class.isAssignableFrom(object.getClass())) {
+        return ((List<Object>) object)
+            .stream()
+            .map(v -> mapValue(descriptor, v))
+            .collect(Collectors.toList());
+      } else {
+        // Create new list in case when originally scala field changed to repeated
+        return Collections.singletonList(mapValue(descriptor, object));
```

Comment:
this looks fragile, can you please elaborate little bit about this edge case?

Reply:
There is a typo in word "scala", it should be scalar (fixed). Basically protobuf allows you to change field name from for example `string field` to `repeated string field` (same for opposite direction) but you will have value stored in original form. In than case we need to create an list from scalar value (to be able to call addRepeated method on builder). For details please see `testCreateProtoWhereOptionalFieldChangedToRepeated` and `testCreateProtoWhereRepeatedFieldChangedToOptional` in `ProtoMessageValueAccessorTest`

Reply:
👍 Ok, now it makes more sense :) Maybe we could change `List.isAssignable..` to `Collection.isAssignable...` to relax the condition?